rs/nervous_system/initial_supply/src/lib.rs

```diff
@@ -0,0 +1,257 @@
+                .ok_or_else(|| "Transaction count overflowed u64.".to_string())?;
+        }
+    }
+
+    // Convert to return type.
+    let result = u64::try_from(result.0).map_err(|err| {
+        format!(
+            "Failed to convert initial supply in e8s does to u64. Reason: {:?}",
+            err,
+        )
+    })?;
+
+    Ok(result)
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct InitialSupplyOptions {
+    /// Give up if more than this many transactions need to be scanned to come up with a result.
+    pub max_transactions: u64,
+
+    /// How many transactions to fetch at a time.
+    pub batch_size: u64,
+}
+
+impl InitialSupplyOptions {
+    /// Sensible values.
+    pub fn new() -> Self {
+        Self {
+            max_transactions: 100_000,
+            batch_size: 250,
+        }
+    }
+}
+
+impl Default for InitialSupplyOptions {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+// TODO: Might probably make sense to spin this in its own library.
+/// A ledger canister does not hold very old transactions. Instead, it passes
+/// old transactions to archive canister(s). Therefore, it is not enough to
+/// simply call a ledger canister's get_transactions method. The response can
+/// indicate that older records must be fetched from an archive canister.
+/// Following those redirects is our main job here.
+#[derive(Debug)]
+struct ThickLedgerClient {
+    ledger_canister_id: CanisterId,
+}
+
+impl ThickLedgerClient {
+    pub fn new(ledger_canister_id: CanisterId) -> Self {
+        Self { ledger_canister_id }
+    }
+
+    pub async fn get_transactions<MyRuntime: Runtime>(
+        &self,
+        start: u64,
+        length: u64,
+    ) -> Result<Vec<Transaction>, String> {
+        let start = Nat::from(start);
+        let length = Nat::from(length);
+        let request = GetTransactionsRequest {
+            start: start.clone(),
+            length: length.clone(),
+        };
+
+        let (mut response,): (GetTransactionsResponse,) =
+            MyRuntime::call_with_cleanup(self.ledger_canister_id, "get_transactions", (request,))
+                .await
+                .map_err(|err| format!("Failed to call ledger: {:?}", err))?;
+
+        normalize_get_transactions_response(start, &mut response)?;
+
+        let mut result = vec![];
+        // Fetch transactions from archive.
+        for archived_range in response.archived_transactions {
+            let mut transactions = self
+                .follow_get_transactions_redirect::<MyRuntime>(archived_range)
+                .await
+                .map_err(|err| {
+                    format!(
+                        "Ledger {} (partially) forwarded us (presumably to archive), \
+                         but that failed: {}",
+                        self.ledger_canister_id, err,
+                    )
+                })?;
+            result.append(&mut transactions);
+        }
+
+        result.append(&mut response.transactions);
```

Comment:
`response.transactions` will only contain [up to 2000 transactions](https://sourcegraph.com/github.com/dfinity/ic@3e4b6640918dcb3d6ff3de129fc019b1ebfd320d/-/blob/rs/rosetta-api/icrc1/ledger/src/lib.rs?L61) here also. However, [the default settings for (SNS) ledgers](https://sourcegraph.com/github.com/dfinity/ic@3e4b6640918dcb3d6ff3de129fc019b1ebfd320d/-/blob/rs/sns/init/src/lib.rs?L609-610) (this applies to ICP and ck-ledgers also) is to only keep 2000 transactions locally (after that, 1000 of them are shipped off to the archive), so in practice there shouldn't be a need to call `get_transactions` on the ledger multiple times to retrieve all the transactions, but you could add a check and potential loop in case some ledgers have a non-standard configuration.